### DISCUSSION TITLE
First-class support for Graphql

### DISCUSSION THREAD

**Discussion by cloudcompute** [directus/directus#12913] [Feature Requests]
Directus is a wonderful platform. But it has one thing missing.. Its support for ```Graphql``` is not powerful which is simply a wrapper over ```Rest```.

Would you guys consider introducing an independent graphql server that lets us easily extend its capabilities with the plugins.. kindly have a look at this wonderful package [```graphql-yoga```](https://www.graphql-yoga.com/)

Two of its features that I like the most:
- Run ```GraphQL Subscriptions``` over HTTP instead of WebSockets via ```Server Sent Events```.
- Extensible GraphQL Engine powered by [```Envelop```](https://www.graphql-yoga.com/docs/features/envelop-plugins)

I am sure it will make Directus much powerful than it is now.

A ```side note```

- For ```gatsby```, there exists a plugin that allows us to extract the content from ```directus``` using simple ```gql```. But there does not exist its equivalent for the all-popular ```Next.js```.

- Using ```directus-sdk``` in ```next``` is what I don't feel like using. I have specified its major limitation [here](https://github.com/directus/directus/discussions/12912)

A ```question```, can you pl. answer.

In order to to pull the data from ```directus``` do we need to write a plugin that lets us use ```gql``` in Next ? Or, could this be achieved in this way.

- Use a graphql-client library like ```apollo``` or ```graphql-request```
 - call the Directus graphql endpoint.. POST /graphql/system _authenticate_
 - After successful authentication, call POST /graphql/system _fetch data from directus_

**Comment by rijkvanzanten** [comment]
> Its support for Graphql is not powerful which is simply a wrapper over Rest.

What do you mean? Directus' GraphQL implementation isn't a wrapper around the rest endpoints at all 🤔 

> Would you guys consider introducing an independent graphql server that lets us easily extend its capabilities with the plugins.. kindly have a look at this wonderful package [graphql-yoga](https://www.graphql-yoga.com/)

Yup! See https://github.com/directus/directus/discussions/4448 for the discussion on that

> Run GraphQL Subscriptions over HTTP instead of WebSockets via Server Sent Events.

Tracking that under https://github.com/directus/directus/discussions/3221, https://github.com/directus/directus/discussions/3255

> In order to to pull the data from directus do we need to write a plugin that lets us use gql in Next ? Or, could this be achieved in this way.

The graphql endpoint is just that: a graphql endpoint. Any graphql compatible client library should work with the directus api as well 👍🏻

**Comment by cloudcompute** [comment]
> Directus' GraphQL implementation isn't a wrapper around the rest endpoints
I read it somewhere in the documentation.
